watcher.go

```diff
@@ -0,0 +1,119 @@
+package main
+
+import (
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/go-logr/logr"
+)
+
+type telegrafClassesOnChange func()
+
+// telegrafClassesWatcher allows monitoring a directory with telegraf classes using
+// fsnotify package and batching multiple events to reduce number of Kubernetes API calls.
+type telegrafClassesWatcher struct {
+	logger        logr.Logger
+	watcherEvents chan fsnotify.Event
+	onChange      telegrafClassesOnChange
+
+	eventCount   uint64
+	eventChannel chan bool
+	eventDelay   time.Duration
+}
+
+// newTelegrafClassesWatcher creates a new instance of telegrafClassesWatcher.
+func newTelegrafClassesWatcher(logger logr.Logger, telegrafClassesDirectory string, onChange telegrafClassesOnChange) (*telegrafClassesWatcher, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	// watching the contents of classes directory requires adding the directory as well as most child elements
+	logger.Info("adding directory to watcher", "directory", telegrafClassesDirectory)
+	watcher.Add(telegrafClassesDirectory)
+
+	items, err := ioutil.ReadDir(telegrafClassesDirectory)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, item := range items {
+		name := item.Name()
+
+		// add all items in the classes directory except for current/previous secret contents that begin with "..", "." and ".."
+		// explicitly add "..data" directory as this is the directory that maps current state of the secret
+		if name == "..data" || (name != "." && !strings.HasPrefix(name, "..")) {
+			p := filepath.Join(telegrafClassesDirectory, name)
+			logger.Info("adding item to watch", "path", p)
+			err = watcher.Add(p)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	w := &telegrafClassesWatcher{
+		watcherEvents: watcher.Events,
+		logger:        logger,
+		onChange:      onChange,
+
+		// allow large number of messages in the channel to avoid blocking
+		eventChannel: make(chan bool, 100),
+
+		// delay by 10 seconds to group multiple fsnotify events into single invocation of callback
+		eventDelay: 10 * time.Second,
+	}
+
+	w.createGoroutines()
+
+	return w, nil
+}
+
+// createGoroutines runs all required goroutines for the watcher.
+func (w *telegrafClassesWatcher) createGoroutines() {
+	go w.batchChanges()
+	go w.monitorForChanges()
+}
+
+// batchChanges is a goroutine that batches invocations of onChange()
+// based on events sent from monitorForChanges().
+func (w *telegrafClassesWatcher) batchChanges() {
+	var previousEventCount uint64
+	for {
+		<-w.eventChannel
```

Comment:
I understand that this goroutine never exits and that it may be acceptable since it likely runs for the entire duration of the program.

However, just an example/suggestion: if you rewrite the iteration using the range loop, you can then close the channels during cleanup and avoid leaking goroutines.

```suggestion
	for range w.eventChannel {
```